README.md
=====
cmsplugin-rt
=====

This package contains a number of basic plugins to kick start your Django-CMS project.
Some default to the Twitter Bootstrap look, but can be adjusted using the setting RT_FRONT_END_FRAMEWORK
Currently "BOOTSTRAP" and "JQUERY-MOBILE" are recognised.
I use this in conjunction with cmsapp-rt, which gives me a global placeholder for the navbar and
banner pictures, as well as some logic for body padding with a fixed-to-top navbar.

button                     a Twitter Bootstrap or JQueryMobile button, with html special characters allowed
text_minimal_markup        plain text with an h2 heading, with links and email addresses activated and html special codes allowed, e.g. &copy;
resizeable_picture         a subclass of cms.plugin.picture which allows absolute or % width and height (sorry about the spelling mistake here!)
navbar                     a Twitter Bootstrap navbar, with self-calculating links to the home page's children, and quick links for admins
self_calc_pagination       shows pagination links where pages after 1 are all the children of page 1 (tailored to Twitter Bootstrap)
hbar                       a simple html <hr> element
spacer                     adds a .spacer div, which I style as clear:both, to switch off float:left and float:right

facebook_button            lots of options allowed
twitter_button             lots of options allowed
mailchimp_form             an inline mailchimp form (tailored to Twitter Bootstrap)
google_font                link to Google fonts (use them via the style modifier plugin)
google_analytics           include tracking code for Google analytics
button_appstore            a simple Download on the Appstore button using code from http://linkmaker.itunes.apple.com/

style_modifier             a plugin that lets the admin change the css directly
                           (a number of Twitter Bootstrap and JQueryMobile classes are pre-configured)
                           add RT_MORE_STYLE_CLASSES to settings.py to add more options to the drop-down list

meta_icons                 include favicon and Apple touch icons in your header
open_graph                 include open-graph tags in your header
rt_carousel                displays a configurable Bootstrap carousel of a selected group of objects
                           you will need to add a line like this to settings.py:
                               ALLOWED_MODELS_IN_RT_CAROUSEL = [{'app_label':'myapp', 'model':'screenshotgroup'},]
                           additional explanation in the comments and at:
                               http://stackoverflow.com/questions/16999421/how-do-i-adapt-django-foreignkey-to-a-reusable-app
                           (rt_carousel is under development and still subject to significant change) 

Detailed documentation is in the "docs" directory (pending).

Requirements
--------------

I built these using:

* Django 1.4
* Django-CMS 2.3.5

I am using them with

* Django 1.5
* Django-CMS 2.4.1

Quick start
-----------

1. Add the desired plugins from cmsplugin-rt to your INSTALLED_APPS setting like this::

      INSTALLED_APPS = (
          ...
          'cmsplugin_rt.button',
          'cmsplugin_rt.facebook_button',
          'cmsplugin_rt.button_appstore',
          'cmsplugin_rt.google_analytics',
          'cmsplugin_rt.google_font',
          'cmsplugin_rt.hbar',
          'cmsplugin_rt.mailchimp_form',
          'cmsplugin_rt.meta_icons',
          'cmsplugin_rt.navbar',
          'cmsplugin_rt.open_graph',
          'cmsplugin_rt.resizeable_picture',
          'cmsplugin_rt.self_calc_pagination',
          'cmsplugin_rt.spacer',
          'cmsplugin_rt.style_modifier',
          'cmsplugin_rt.text_minimal_markup',
          'cmsplugin_rt.twitter_button',
      )

2. If you are using JQueryMobile, add to settings.py:

      RT_FRONT_END_FRAMEWORK = "JQUERY-MOBILE"  # "BOOTSTRAP" is the default; only these two are known

   You can also set this to "" to remove Style Modifier's default Bootstrap class options.

3. To add custom classes to the Style Modifier, add to settings.py something like this:

      from django.utils.translation import ugettext_lazy as _
      RT_MORE_STYLE_CLASSES = ((".banner", _("banner")),
                     (".warning", _("warning text")),      # example classes only
                     )

4. The button template is configured for Twitter Bootstrap, although the admin panel shows choices appropriate for
   JQueryMobile as well.
   If you want to use buttons with JQueryMobile, override the template by adding to your own project the file:
      templates/button_plugin.html, containing:

      {% load allow_special %}
      <a data-role="button" {% ifequal instance.button_type "inline" %}data-inline="true"{% endifequal %}
            {% ifequal instance.button_size "btn-mini" %}data-mini="true"{% endifequal %} data-ajax="false" href="{{ link }}">
        {{ instance.button_text|allow_special }}
        {% if instance.arrows %} &raquo; {% endif %}
      </a>

5. Run `python manage.py syncdb` (or use `python manage.py migrate` if you are using South) to create the models.

6. Add the plugins to your CMS pages in the admin panel.

7. Some plugins use 'spacer', 'align-center', 'align-left', 'align-right' classes, which you should implement in your css, e.g.:
    .align-right {
	    float: right;
    }
    .align-left {
	    float: left;
    }
    .align-center {
	    margin-left:auto;
	    margin-right:auto;
	    text-align:center;
	    display: block;  /* text-align:center and display:block are critical to getting this to work */
    }
    .spacer {
	    clear: both;
    }



CHANGES
0.1.1
-----
Changed URL in setup.py in an effort to fix a problem with pip install cmsplugin-rt

0.1.2
-----
Added mailchimp_form
Added zip_safe=False to setup

0.1.3
-----
Fixed URL of open graph image meta tag so it is absolute, not relative

0.1.4
-----
Added navbar plugin
Added a tests directory, not functional yet (can you help me fix it?)

0.1.5
-----
Added google_fonts and google_analytics plugins
Revised style_modifier to include font_family

0.1.6
-----
Added p to style modifier
Added css width and height to resizeable picture as well as img properties

0.2.0
-----
Fixed South migration files to use the correct plugin names (as I changed their names earlier without doing this)

0.3.0
-----
Added support for JQueryMobile in StyleModifier plugin.

0.3.1
-----
Added URL for github repo to setup.py

0.3.2
-----
Added desirable css to README
Added max-height and max-width to resizeable_picture plugin

0.3.3
-----
Added id navbar to navbar template, for easier styling
Defaults meta-icon apple-touch icon to fav-icon if missing

0.3.4
-----
Added id navbar to style modifier's navbar entry

0.3.5
-----
Changed style modifier primary button class to .btn.btn-primary (from .btn-primary, which often didn't work as intended)

0.3.6
-----
Changed style modifier - fixed 126-char class name on .btn.btn-primary; and removed background-repeat:no-repeat

0.3.7
-----
Changed google font so it does not assume http

0.3.8
-----
Changed allow_special tag so that links with http(s):// do not also need www to be turned into links

0.3.9
-----
Fixed small error in prev change

0.4.0
-----
Navbar does not display brand if empty (improves display on devices)

0.4.1
-----
Google requests that the analytics code be in the head section, so do not let sekizai move it to the end

0.4.2
-----
Added Download on the App Store button

0.4.3
-----
Made button usable in the text plugin

0.4.4
-----
Allowed >, ", ' in some style modifier fields. Remove all <, script and expression.

0.4.5
-----
Added active and hover modifiers for navbar links

0.4.6
-----
Added draft version of rt_carousel

0.4.7
-----
Changed README.md to README.rst
Removed suffixes from CHANGES.txt and LICENSE.txt

0.4.8
-----
Changed README formatting

0.4.9
-----
If only one item in the carousel, do not show fwd and back buttons

0.5.0
-----
Fix duplication problem with button links in admin, as per link plugin

0.5.1
-----
Allow button link to be 240 chars (was 80)


cmsplugin_rt/button/models.py
# Plugin models

from cms.models.pluginmodel import CMSPlugin
from cms.models.pagemodel import Page
from django.utils.translation import ugettext_lazy as _
from django.db import models
from django.conf import settings

class ButtonPluginModel(CMSPlugin):
    front_end = getattr(settings,'RT_FRONT_END_FRAMEWORK','BOOTSTRAP').upper()

    if (front_end=="BOOTSTRAP"):
        CLASS_CHOICES =(("", _("default")),
                    ("btn-primary", _("primary")),
                    ("btn-info", _("info")),
                    ("btn-success", _("success")),
                    ("btn-warning", _("warning")),
                    ("btn-danger", _("danger")),
                    ("btn-inverse", _("inverse")),
                    ("btn-link", _("link")),
        )
        SIZE_CHOICES = (("", _("default")),
                    ("btn-large", _("large")),
                    ("btn-small", _("small")),
                    ("btn-mini", _("mini")),
        )
    elif (front_end=="JQUERY-MOBILE"):
        CLASS_CHOICES =(("", _("default")),
                    ("inline", _("inline")),
        )
        SIZE_CHOICES = (("", _("default")),
                    ("btn-mini", _("mini")),
        )
    else:
        CLASS_CHOICES = ()
        SIZE_CHOICES  = ()

    button_type = models.CharField(_("button type"), max_length=16, blank=True, choices=CLASS_CHOICES)
    button_size = models.CharField(_("button size"), max_length=16, blank=True, choices=SIZE_CHOICES)
    button_link = models.CharField(max_length=240, default='', blank=True)
    page_link = models.ForeignKey(Page, verbose_name=_("page"), blank=True, null=True, help_text=_("A link to a page overrides the above button link."))
    button_text = models.CharField(max_length=80, default='Click here', help_text=_("HTML symbol codes are allowed, e.g. &amp;hearts; for &hearts;."))
    arrows = models.BooleanField()
    
    def __unicode__(self):
        return self.button_text

    search_fields = ('button_text','button_link',)

